File: web-cat-src/Utilities/ElectronicSubmitter/src/org/webcat/submitter/internal/protocols/FileProtocol.java
/*==========================================================================*\
 |  $Id: FileProtocol.java,v 1.2 2010/09/14 18:13:30 aallowat Exp $
 |*-------------------------------------------------------------------------*|
 |  Copyright (C) 2006-2009 Virginia Tech
 |
 |  This file is part of Web-CAT Electronic Submitter.
 |
 |  Web-CAT is free software; you can redistribute it and/or modify
 |  it under the terms of the GNU General Public License as published by
 |  the Free Software Foundation; either version 2 of the License, or
 |  (at your option) any later version.
 |
 |  Web-CAT is distributed in the hope that it will be useful,
 |  but WITHOUT ANY WARRANTY; without even the implied warranty of
 |  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 |  GNU General Public License for more details.
 |
 |  You should have received a copy of the GNU General Public License along
 |  with Web-CAT; if not, see <http://www.gnu.org/licenses/>.
\*==========================================================================*/

package org.webcat.submitter.internal.protocols;

import java.io.FileOutputStream;
import java.io.IOException;
import java.net.URL;
import java.net.URLDecoder;
import org.webcat.submitter.ILongRunningTask;
import org.webcat.submitter.IProtocol;
import org.webcat.submitter.SubmissionManifest;

//--------------------------------------------------------------------------
/**
 * A protocol for the "file" URI scheme that supports writing the submitted file
 * to a location on the local file system (or, perhaps, to a mounted network
 * location).
 *
 * @author  Tony Allevato (Virginia Tech Computer Science)
 * @author  latest changes by: $Author: aallowat $
 * @version $Revision: 1.2 $ $Date: 2010/09/14 18:13:30 $
 */
public class FileProtocol implements IProtocol
{
    //~ Methods ...............................................................

    // ----------------------------------------------------------
    /**
     * @see IProtocol#submit(SubmissionManifest, ILongRunningTask)
     */
    public void submit(SubmissionManifest manifest, ILongRunningTask task)
    throws IOException
    {
        URL url = manifest.getResolvedTransport(null).toURL();

        String path = url.getFile();
        path = URLDecoder.decode(path, "utf-8");

        FileOutputStream outStream = new FileOutputStream(path);

        manifest.packageContentsIntoStream(outStream, task, null);

        outStream.close();
    }


    // ----------------------------------------------------------
    /**
     * @see IProtocol#hasResponse()
     */
    public boolean hasResponse()
    {
        return false;
    }


    // ----------------------------------------------------------
    /**
     * @see IProtocol#getResponse()
     */
    public String getResponse()
    {
        return null;
    }
}

File: web-cat-src/VisualStudio/WebCATSubmitter/WebCATSubmitterCore/Internal/Protocols/FileProtocol.cs
/*==========================================================================*\
 |  $Id: FileProtocol.cs,v 1.2 2008/12/12 01:41:40 aallowat Exp $
 |*-------------------------------------------------------------------------*|
 |  Copyright (C) 2008 Virginia Tech
 |
 |  This file is part of the Web-CAT Electronic Submission engine for the
 |	.NET framework.
 |
 |  Web-CAT is free software; you can redistribute it and/or modify
 |  it under the terms of the GNU Affero General Public License as published
 |  by the Free Software Foundation; either version 3 of the License, or
 |  (at your option) any later version.
 |
 |  Web-CAT is distributed in the hope that it will be useful,
 |  but WITHOUT ANY WARRANTY; without even the implied warranty of
 |  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 |  GNU General Public License for more details.
 |
 |  You should have received a copy of the GNU Affero General Public License
 |  along with Web-CAT; if not, see <http://www.gnu.org/licenses/>.
\*==========================================================================*/

using System;
using System.Collections.Generic;
using System.Text;
using System.IO;
using System.Net;

namespace WebCAT.Submitter.Internal.Protocols
{
	/// <summary>
	/// An implementation of IProtocol that permits submitting a file to a
	/// location on the file system (either local or a networked drive).
	/// </summary>
	internal class FileProtocol : IProtocol
	{
		//  -------------------------------------------------------------------
		public void Submit(SubmissionManifest manifest)
		{
			WebRequest req = WebRequest.Create(manifest.ResolvedTransport);

			if (req is FileWebRequest)
			{
				FileWebRequest request = (FileWebRequest)req;

				request.Method = WebRequestMethods.File.UploadFile;

				Stream stream = request.GetRequestStream();
				manifest.PackageContentsIntoStream(stream);
				stream.Close();
			}
		}

		//  -------------------------------------------------------------------
		/// <summary>
		/// Returns false since file: protocol submissions do not generate a
		/// response.
		/// </summary>
		public bool HasResponse
		{
			get
			{
				return false;
			}
		}

		//  -------------------------------------------------------------------
		/// <summary>
		/// Returns null since file: protocol submissions do not generate a
		/// response.
		/// </summary>
		public string SubmissionResponse
		{
			get
			{
				return null;
			}
		}
	}
}
